walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,95 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
+
+  /** 
+   * Return all possible time ranges to hold a meeting of specific duration.
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Find which times work for mandatory attendees
+    ArrayList<TimeRange> busyRangesForMandatoryAttendees = findBusyRanges(request.getAttendees(), events);
+    Collections.sort(busyRangesForMandatoryAttendees, TimeRange.ORDER_BY_START);
+    Collection<TimeRange> freeRangesForMandatoryAttendees = findFreeRanges(request, busyRangesForMandatoryAttendees);
+    
+    // Factor in optional attendees if there are any
+    if (!request.getOptionalAttendees().isEmpty()) {
+      ArrayList<TimeRange> busyRangesForOptionalAttendees = findBusyRanges(request.getOptionalAttendees(), events);
+      Collections.sort(busyRangesForOptionalAttendees, TimeRange.ORDER_BY_START);
+      
+      ArrayList<TimeRange> busyRangesForAllAttendees = new ArrayList<>();
+      busyRangesForAllAttendees.addAll(busyRangesForMandatoryAttendees);
+      busyRangesForAllAttendees.addAll(busyRangesForOptionalAttendees);
+      Collections.sort(busyRangesForAllAttendees, TimeRange.ORDER_BY_START);
+      Collection<TimeRange> freeRangesForAllAttendees = findFreeRanges(request, busyRangesForAllAttendees);
```

Comment:
This could probably be a helper method. It would increase the clarity of the code.